projects/project-2/readings/Project 2 - SPARQL library

```diff
@@ -0,0 +1,297 @@
+Detecting Properties Shared Across Multiple CCO Classes Level 3 - 20 points
+Detecting Inconsistent Property Ranges Across Subclass Hierarchies Level 3 - 20 points
+
+1. Detect Classes Without Any Instances (Level 8 - 0 points)
+
+Title: Empty Class Detection
+Constraint Description: Identifies classes that have no instances associated with them.
+Severity: Warning
+
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+SELECT ?class
+WHERE {
+  ?class a rdfs:Class .
+  FILTER NOT EXISTS {
+    ?instance a ?class .
+  }
+}
+
+Purpose: ​This query retrieves all classes that do not have any instances associated with them.​ 
+It uses FILTER NOT EXISTS to ensure that only classes without any defined instances are selected,helping to identify potentially unused or unnecessary classes in the ontology.
+
+2. Identifying Disjoint Classes with Common Instances
+
+Title: Disjoint Class Violation
+Constraint Description: Detects instances that belong to classes declared as disjoint.
+Severity: Error
+
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+SELECT ?class1 ?class2 ?instance
+WHERE {
+  ?instance a ?class1 .
+  ?instance a ?class2 .
+  ?class1 owl:disjointWith ?class2 .
+}
+
+The primary intention of this query is to identify pairs of classes that are designated as disjoint but have instances that belong to both classes.​ 
+This situation indicates a potential inconsistency within the ontology since disjoint classes should not share any instances.
+
+Running this query enables to detect potential issues in the class hierarchy and relationships, ensuring the integrity of the ontology is maintained. 
+
+3. Detecting Unused Object Properties (level 7 - 2 points)
+
+Title: Unused Object Property Detection
+Constraint Description: Identifies object properties that are defined but not used in any instances.
+Severity: Warning
+
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+SELECT ?property
+WHERE {
+  ?property a owl:ObjectProperty .
+  FILTER NOT EXISTS {
+    ?subject ?property ?object .
+  }
+}
+
+​The aim of this query is to identify object properties within the ontology that are defined but not utilized in any instances.​ 
+Detecting such unused properties helps maintain a clean and efficient ontology by identifying potentially redundant definitions.
+
+4. Finding Classes with Conflicting Disjointness and Subclass Relationships
+
+Title: Disjoint Subclass Conflict
+Constraint Description: Identifies classes that are marked as disjoint but share a common superclass.
+Severity: Error
+
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+SELECT ?class1 ?class2
+WHERE {
+  ?class1 owl:disjointWith ?class2 .
+  ?class1 rdfs:subClassOf ?commonSuperclass .
+  ?class2 rdfs:subClassOf ?commonSuperclass .
+}
+
```

Comment:
Maybe it will be helpful to to consider subclasses at multiple levels of hierarchy.. so you can use rdfs:subClassOf+

SELECT ?class1 ?class2
WHERE {
  ?class1 owl:disjointWith ?class2 .
  ?class1 rdfs:subClassOf+ ?commonSuperclass .
  ?class2 rdfs:subClassOf+ ?commonSuperclass .
}

Hope this helps!

Reply:
Thanks!